### ISSUE TITLE
Show current filters during workflow submission & allow changes

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-web#442]
The workflow submission form should show what filters are currently set. 

As a user, I can see what filters would be used for a workflow submission.
(potentially I can also see what images would then be used for processing)

As a user, I can change what filters will be set during workflow submission.

I haven't figured out the UI aspects for this. Is this currently feasible from a backend perspective @tcompa ?

**Comment by tcompa** [comment]
This issue should be tackled after #440, which already changes the submission modal quite a bit.

My first guess is that we would make the submission modal a "multi-step" flow. Right now (independently on #440) it already has two steps, but the second one is a dummy one for the confirmation.

We could move towards a three-steps flow:

A. The current submission modal
B. The new filter feature
C. The confirmation

> I haven't figured out the UI aspects for this. Is this currently feasible from a backend perspective @tcompa ?

Two possible designs are already supported, while the third below is not.

One supported option:

1. Upon submission (as part of step B above), the user can see the current filters for the dataset and for the first selected workflow task. If needed, they can send a PATCH-dataset request to modify the dataset filters before submission (with the same UI as in the dataset page).
2. As a more advanced feature (still doable, but let's postpone it for now), we could link to a custom image-list page that shows only the images that would be used in the **first** task to be executed
3. Depending on where/how we tackle #463, this feature can then be made arbitrarily more flexible. Again, I suggest we stick with the basic version for now.

Another supported option:

4. Upon submission (as part of step B above), the user can see the current filters for the dataset and for the first selected workflow task. If needed, they can send a PATCH-workflowtask request to modify the workflowtask input_filters before submission (with the same UI as in the workflowtask input-filters page).
5. Same as 2.
6. Same as 3.

What is not supported in the backend is the following:

7. Upon submission (as part of step B above), the user can see the current filters for the dataset and for the first selected workflow task. If needed, they can set additional filters which are associated to this specific job and take priority over any other existing filter (the dataset and workflow-task ones).

**Comment by jluethi** [comment]
> the user can see the current filters for the dataset and for the first selected workflow task.

Agreed that it's just for the first selected workflow task!

We only need to show this more complex dialogue in the case of "Continue Workflow", because Run workflow & restart workflow both will have empty image lists.
(there could be edge-cases where a users adds things to the image list before clicking run workflow, but not for restart workflow. In any case, the feature is most important for Continue workflow)


For the continue workflow task, having this new suggested flow sounds reasonable:

> A. The current submission modal
> B. The new filter feature
> C. The confirmation

As to whether the user would want to change dataset filters or workflow task filters: Primarily the dataset filters, as they would require switching the page otherwise. But I haven't thought deeply enough about the interaction of the 2 filters and e.g. a user having set a workflow task filter that they are actually not happy about when they see it applied etc.

I agree that there shouldn't be job filters, the filter system is already complex enough! But we'll need to think through the scenarios for how filters are set, how we display this to the user during submission & what type of editing can reasonably be exposed in the submission modal

**Comment by jluethi** [comment]
First thoughts on what should be shown in the modal:
We should show a list of all the currently applied filters (dataset & workflow task) which can be modified. And if we can have a button that then shows "show me which images actually would be selected" (e.g. a version of the dataset image list with all dataset filters applied + workflow task filters), that would be an upgrade!

I'll try to sketch some options for this later to make it more visual

**Comment by zonia3000** [comment]
Is this as expected?

![image](https://github.com/fractal-analytics-platform/fractal-web/assets/5459334/fd9710d6-61f5-4fd5-a496-a4259c440d38)

![image](https://github.com/fractal-analytics-platform/fractal-web/assets/5459334/2a55ba5a-7465-4c8a-a653-83e8713b97df)

![image](https://github.com/fractal-analytics-platform/fractal-web/assets/5459334/1f686c84-1fde-4fa1-bff0-9882a4070d5e)

**Comment by jluethi** [comment]
Functionally, I think that's a good start.

Layout & user-flow wise, I have a few comments, plus 1 functionality question.

Page 1: Let's have 2 buttons:
a) Edit Filters => sends you to the edit filters modal (Page 2)
b) Run => Directly sends you to Page 3

Page 2:
Add text:
"Currently, the following filters are applied to the image list before it is passed to the first task:"

I think we should not allow people to edit the names of the attributes & types there. e.g. if I have an attribute `is_3D`, then I can change the filter from True to False, but not rename the attribut/filter.

In the future (not just yet): We add a button for "show filtered image list" (maybe also to page 3)

Page 3:
Separate main options from filters, e.g. with a line.
Add a "Applied filters" header
Make the headers smaller for attributes & types.

Functionality questions
Affecting page 2: What actually happens when a user edits filters on page 2 above? Is the actual filter edited in the dataset or the workflow filter? [that would be my expectations, but want to verify] What happens to new filters that are added? Are the added to the workflow task or the dataset

**Comment by zonia3000** [comment]
> I think we should not allow people to edit the names of the attributes & types there. e.g. if I have an attribute `is_3D`, then I can change the filter from True to False, but not rename the attribut/filter.

So in this case I also have to remove the buttons "Add attribute filter" and "Add type filter", the buttons that deletes the attributes/filters and the dropdown that allow to change the type of an attribute?
 
> Functionality questions Affecting page 2: What actually happens when a user edits filters on page 2 above? Is the actual filter edited in the dataset or the workflow filter? [that would be my expectations, but want to verify] What happens to new filters that are added? Are the added to the workflow task or the dataset

I'm currently applying the changes to the dataset endpoint.

**Comment by jluethi** [comment]
> So in this case I also have to remove the buttons "Add attribute filter" and "Add type filter", the buttons that deletes the attributes/filters and the dropdown that allow to change the type of an attribute?

No, I would allow users to add new filters or delete filters. It's just confusing if they rename existing filters from that interface.
Given that existing filters aren't typically set in the run dialogue, this page is primarily about:
a) Showing what filters are set
b) allowing users to change the value of the filters
c) (rarer) add new filters)
d) (rarer) delete filters that are currently applied

Renaming existing filters would be something I think a user should do on the dataset page, not during workflow submission

**Comment by tcompa** [comment]
I think some clarifications are needed here.

Let's stick with the modal structure that is being discussed above: step 1 for general options, step 2 (optional) concerns filters, step 3 is the submission confirmation.

The question here is what goes in step 2

# Option 0

Let's start with a read-only version of step, so that we improve the specifications. Note that within this option the presence of both steps 2 and 3 becomes redundant.

When showing filters, in step 2, we are combining two filter sets:
* `workflow.task_list[index].input_filters`, where `index` is the positional index of the selected first task (or 0, by default)
* `dataset.filters`, where `dataset` is the selected one (that is, the one that will be used for the submission)

This means that there should be, in principle, four blocks: wftask/types, wftask/attributes, dataset/types, dataset/attributes.
This is clearly quite complex, to show, and we could decide to merge these four into two block (types and attributes). The only mitigation that comes to mind is that the wftask block could be hidden when they are unset (which is the most typical situation).

Note that just showing the four blocks is not enough, because some filters would not be used in the first workflowtask. In principle we can include this information, but the risk here is to make a very complex modal.

I am not very convinced of this merge:
1. Dataset filters are used for all tasks of the job, while workflowtask filters are only applied for the first task (and subsequent tasks may actually have their own filters, which will be used by the backend). By merging them together, we are not offering a fully transparent view of what will take place in the backend.
2. The logic of how to merge the two filter sets (dataset and workflowtask) is currently quite simple (the workflowtask ones take priority), but it could become more complex in the future (https://github.com/fractal-analytics-platform/fractal-server/issues/1401). In that case, it should not be implemented both in server and client, but we should consider offering a dedicated endpoint that can be called from the client.

Side note: if we merge the filters and (in a future feature) we also show the filtered image list, we are showing the list of images that would be used for the first task within the runner, but we are potentially **not** showing some of the images that would be used for the second task. If one hits this edge case, then seeing the image list is actually misleading.

# Option 1

Option 0 was all about reading current filters, in order to clarify the situation. What if we also want to expose some editing feature?

The version that I find more clear is the one where all four blocks are displayed and editable: wftask/types, wftask/attributes, dataset/types, dataset/attributes. This UI is quite complex, and requires the user to understand the difference between the two kinds of filters.

Note that if we had merged the two sets of wftask/dataset filters together, then it's unclear what the form would edit/add/remove.
* For the edit feature, we can somehow find out whether this concerns the workflowtask or the dataset.
* For remove, it would be quite hard e.g. when the same key is present both in workflowtask and dataset. Say that removal defaults to workflowtask, then the "remove" action will results in a change of what is displayed -> very confusing.
* For add, it's undefined where we should add a new filter. It's easy to pick a default, but then would the user understand what they are modifying? If so, I'd say it's better to just show all the four blocks.

# High level

I'd hesitant to add too many interactions to this single modal, but if we want these features to all be present then we should find better definition of what actions are allowed.

**Comment by tcompa** [comment]
OK, thanks.
Thus let's start by implementing the read-only version with merged dataset/wftask filters, and re-discuss add/edit/remove later.

**Comment by jluethi** [comment]
Sounds good! And that means we can drop modal 2 for the moment, the confirmation page "just" shows the combined filters.

We can then discuss later on introducing an intermediary modal that allows editing of filters

**Comment by zonia3000** [comment]
These are the read-only filters in the confirmation step of the modal. k1 is an attribute, t1 and t2 are types, merged prioritizing first task filters:

![image](https://github.com/fractal-analytics-platform/fractal-web/assets/5459334/83faeb4e-781c-4520-b313-6d6418d3b3e1)

**Comment by jluethi** [comment]
This looks good to me!
